eva/readers/decord_reader.py

```diff
@@ -71,38 +92,54 @@ def _read(self) -> Iterator[Dict]:
         logger.debug("Reading frames")
 
         if self._sampling_type == IFRAMES:
-            iframes = video.get_key_indices()
+            iframes = v_reader.get_key_indices()
             idx = 0
             for begin, end in range_list:
                 while idx < len(iframes) and iframes[idx] < begin:
                     idx += self._sampling_rate
 
                 while idx < len(iframes) and iframes[idx] <= end:
                     frame_id = iframes[idx]
-                    frame = video[frame_id]
+                    if has_audio:
+                        frame_audio, frame = av_reader[frame_id]
+                        frame_audio = frame_audio.asnumpy()
+                        if self._read_video is None or not self._read_video:
+                            frame = np.empty(0)
+                    else:
+                        frame = v_reader[frame_id]
+                        frame_audio = np.empty(0)
+                    frame = frame.asnumpy()
+                    timestamp = v_reader.get_frame_timestamp(frame_id)[0]
                     idx += self._sampling_rate
                     if frame is not None:
                         yield {
                             "id": frame_id,
                             "data": frame,
-                            "seconds": math.floor(
-                                video.get_frame_timestamp(frame_id)[0]
-                            ),
+                            "seconds": math.floor(timestamp),
+                            "audio": frame_audio,
                         }
                     else:
                         break
         elif self._sampling_rate == 1:
             for begin, end in range_list:
                 frame_id = begin
                 while frame_id <= end:
-                    frame = video[frame_id]
+                    if has_audio:
+                        frame_audio, frame = av_reader[frame_id]
+                        frame_audio = frame_audio.asnumpy()
+                        if self._read_video is None or not self._read_video:
+                            frame = np.empty(0)
+                    else:
+                        frame = v_reader[frame_id]
+                        frame_audio = np.empty(0)
+                    frame = frame.asnumpy()
+                    timestamp = v_reader.get_frame_timestamp(frame_id)[0]
```

Comment:
I see a lot of code duplication. How about we create private functions to improve readability and maintainability?

Reply:
Done